cpp/src/arrow/ArrowConfig.cmake.in

```diff
@@ -87,6 +87,10 @@ if(NOT (TARGET arrow_shared OR TARGET arrow_static))
       set_target_properties(
         arrow_static PROPERTIES INTERFACE_LINK_LIBRARIES
         "${arrow_static_interface_link_libraries};arrow_bundled_dependencies")
+        # Propagate dependencies like OpenSSL
+        set_target_properties(
+          arrow_bundled_dependencies PROPERTIES INTERFACE_LINK_LIBRARIES
+          "${arrow_static_interface_link_libraries}")
```

Comment:
I don't know why this is needed.
Is there a use-case that uses `arrow_bundled_dependencies` without `arrow_static`? 

BTW, indent for this code is broken.

Reply:
Ah, the problem here is the order of libraries to the linker. As described above, currently, the order is OpenSSL, then libarrow_bundled_dependencies, so linking will fail because libarrow_bundled_dependencies itself also depends on OpenSSL. Adding this gets CMake to repeat OpenSSL after libarrow_bundled_dependencies and then the linker will find the necessary symbols.

Reply:
I see. Then we should list OpenSSL in `INTERFACE_LINK_LIBRARIES` of `arrow_bundled_dependencies` not `arrow_static` and `INTERFACE_LINK_LIBRARIES` of `arrow_static` should list only `arrow_bundled_dependencies` something like this:

```diff
diff --git a/cpp/src/arrow/ArrowConfig.cmake.in b/cpp/src/arrow/ArrowConfig.cmake.in
index 6209baeec..ac2b9a027 100644
--- a/cpp/src/arrow/ArrowConfig.cmake.in
+++ b/cpp/src/arrow/ArrowConfig.cmake.in
@@ -38,6 +38,7 @@ set(ARROW_LIBRARY_PATH_SUFFIXES "@ARROW_LIBRARY_PATH_SUFFIXES@")
 set(ARROW_INCLUDE_PATH_SUFFIXES "@ARROW_INCLUDE_PATH_SUFFIXES@")
 set(ARROW_SYSTEM_DEPENDENCIES "@ARROW_SYSTEM_DEPENDENCIES@")
 set(ARROW_BUNDLED_STATIC_LIBS "@ARROW_BUNDLED_STATIC_LIBS@")
+set(ARROW_STATIC_INSTALL_INTERFACE_LIBS "@ARROW_STATIC_INSTALL_INTERFACE_LIBS@")
 
 include("${CMAKE_CURRENT_LIST_DIR}/ArrowOptions.cmake")
 
@@ -79,14 +80,9 @@ if(NOT (TARGET arrow_shared OR TARGET arrow_static))
         PROPERTIES
           IMPORTED_LOCATION
           "${arrow_lib_dir}/${CMAKE_STATIC_LIBRARY_PREFIX}arrow_bundled_dependencies${CMAKE_STATIC_LIBRARY_SUFFIX}"
+          INTERFACE_LINK_LIBRARIES
+          "${ARROW_STATIC_INSTALL_INTERFACE_LIBS}"
         )
-
-      get_property(arrow_static_interface_link_libraries
-                  TARGET arrow_static
-                  PROPERTY INTERFACE_LINK_LIBRARIES)
-      set_target_properties(
-        arrow_static PROPERTIES INTERFACE_LINK_LIBRARIES
-        "${arrow_static_interface_link_libraries};arrow_bundled_dependencies")
     endif()
   endif()
 endif()
diff --git a/cpp/src/arrow/CMakeLists.txt b/cpp/src/arrow/CMakeLists.txt
index 5736c557b..66a24fbfc 100644
--- a/cpp/src/arrow/CMakeLists.txt
+++ b/cpp/src/arrow/CMakeLists.txt
@@ -526,6 +526,9 @@ endif()
 
 if(ARROW_BUILD_BUNDLED_DEPENDENCIES)
   string(APPEND ARROW_PC_LIBS_PRIVATE " -larrow_bundled_dependencies")
+  set(ARROW_STATIC_INSTALL_INTERFACE_LIBS_REAL "arrow_bundled_dependencies")
+else()
+  set(ARROW_STATIC_INSTALL_INTERFACE_LIBS_REAL ${ARROW_STATIC_INSTALL_INTERFACE_LIBS})
 endif()
 # Need -latomic on Raspbian.
 # See also: https://issues.apache.org/jira/browse/ARROW-12860
@@ -557,7 +560,7 @@ add_arrow_lib(arrow
               SHARED_INSTALL_INTERFACE_LIBS
               ${ARROW_SHARED_INSTALL_INTERFACE_LIBS}
               STATIC_INSTALL_INTERFACE_LIBS
-              ${ARROW_STATIC_INSTALL_INTERFACE_LIBS})
+              ${ARROW_STATIC_INSTALL_INTERFACE_LIBS_REAL})
 
 add_dependencies(arrow ${ARROW_LIBRARIES})
 
```